charms/feast-ui/src/charm.py

```diff
@@ -0,0 +1,138 @@
+        self.kubeflow_dashboard_sidebar = KubeflowDashboardLinksRequirer(
+            charm=self,
+            relation_name="dashboard-links",
+            dashboard_links=DASHBOARD_LINKS,
+        )
+
+        # Patch Kubernetes service to expose the HTTP port
+        self.service_patcher = KubernetesServicePatch(
+            self,
+            [ServicePort(HTTP_PORT, name="http")],
+            service_name=f"{self.model.app.name}",
+        )
+
+        self.charm_reconciler = CharmReconciler(self)
+
+        # Leadership gate
+        self.leadership_gate = self.charm_reconciler.add(
+            component=LeadershipGateComponent(charm=self, name="leadership-gate"),
+            depends_on=[],
+        )
+
+        self.ingress_relation = self.charm_reconciler.add(
+            SdiRelationBroadcasterComponent(
+                charm=self,
+                name="relation:ingress",
+                relation_name="ingress",
+                data_to_send={
+                    "prefix": "/feast",
+                    "rewrite": "/",
+                    "service": self.model.app.name,
+                    "namespace": self.model.name,
+                    "port": HTTP_PORT,
+                },
+            ),
+            depends_on=[self.leadership_gate],
+        )
+
+        # Store config from relation
+        self.store_configuration_receiver = self.charm_reconciler.add(
+            component=StoreConfigurationReceiverComponent(charm=self, relation_name=RELATION_NAME),
+            depends_on=[self.leadership_gate],
+        )
+
+        # Pebble service layer
+        self.pebble_service_container = self.charm_reconciler.add(
+            component=FeastUIPebbleService(
+                charm=self,
+                name="feast-ui-pebble-service",
+                container_name=CONTAINER_NAME,
+                service_name=SERVICE_NAME,
+                files_to_push=self._generate_feature_store_file(),
+            ),
+            depends_on=[self.leadership_gate, self.store_configuration_receiver],
+        )
+
+        self.charm_reconciler.install_default_event_handlers()
+
+    def _generate_feature_store_file(self) -> List[ContainerFileTemplate]:
+        """Generate and return the feature_store.yaml to push to the container."""
+        try:
+            yaml_data = self.store_configuration_receiver.component.get_feature_store_yaml()
+        except ErrorWithStatus as err:
+            self.unit.status = err.status
+            logger.info(f"Stopped early during feature store file generation: {err}")
+            return []
+
+        if not yaml_data:
+            self.unit.status = WaitingStatus("feature_store.yaml is missing or empty")
+            return []
+
+        with tempfile.NamedTemporaryFile(delete=False, mode="w", suffix=".yaml") as f:
+            f.write(yaml_data)
+            path = f.name
+
+        return [ContainerFileTemplate(source_template_path=path, destination_path=DEST_PATH)]
```

Comment:
I'd expect this logic to be encapsulated in the `StoreConfigurationReceiverComponent`, with the BaseCharm pattern we should not be setting the status outside of components.

Reply:
Synced with @NohaIhab, we decided to keep it as it is for now. I have filed issue https://github.com/canonical/charmed-kubeflow-chisme/issues/148 describing the problem.